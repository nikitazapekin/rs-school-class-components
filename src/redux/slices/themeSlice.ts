import { createSlice  } from '@reduxjs/toolkit';
import {  UserDataArray } from '../types';
 
export interface UserData {
	total_count: number;
	incomplete_results: boolean;
	items: UserDataArray;
}
 
interface ThemeSlice {
    isDark: boolean
}
  

  
const initialState: ThemeSlice = {
 isDark: false
   
};
const themeSlice = createSlice({
	name: 'app',
	initialState,
	reducers: {
		setTheme(state) {
			state.isDark = !state.isDark
		},
	 
	},

 

});
export const {
setTheme
} = themeSlice.actions;
export default themeSlice.reducer;
