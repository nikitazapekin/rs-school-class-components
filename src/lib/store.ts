//lib/store.ts

import { combineReducers, configureStore } from '@reduxjs/toolkit';
import type { ThunkAction, Action } from '@reduxjs/toolkit';
import appSlice from './slices/appSlice';
import { setupListeners } from '@reduxjs/toolkit/query';
import { githubApi } from './slices/querySlice';
import { usersGithubApi } from './slices/userQuerySlice';
import selectedElementsSlice from './slices/selectedElementsSlice';
import testSlice from './slices/testSlice';

const rootReducer = combineReducers({
  appSlice: appSlice,
  selectedElementsSlice: selectedElementsSlice,
  testSlice: testSlice,
  [githubApi.reducerPath]: githubApi.reducer,
  [usersGithubApi.reducerPath]: usersGithubApi.reducer,
});

export const makeStore = () => {
  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(githubApi.middleware).concat(usersGithubApi.middleware),
  });
};

// Infer the type of makeStore
export type AppStore = ReturnType<typeof makeStore>;
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<AppStore['getState']>;
export type AppDispatch = AppStore['dispatch'];

// Set up listeners
const store = makeStore();
setupListeners(store.dispatch);

export default store;

/*import { configureStore } from '@reduxjs/toolkit'

export const makeStore = () => {
  return configureStore({
    reducer: {},
  })
}

// Infer the type of makeStore
export type AppStore = ReturnType<typeof makeStore>
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<AppStore['getState']>
export type AppDispatch = AppStore['dispatch']
*/




/*
import { combineReducers, configureStore } from '@reduxjs/toolkit';
import type { ThunkAction, Action } from '@reduxjs/toolkit';
import appSlice from './slices/appSlice';
import { setupListeners } from '@reduxjs/toolkit/query';
import { githubApi } from './slices/querySlice';
import { usersGithubApi } from './slices/userQuerySlice';
import selectedElementsSlice from './slices/selectedElementsSlice';

export const rootReducer = combineReducers({
	appSlice: appSlice,
	selectedElementsSlice: selectedElementsSlice,
	[githubApi.reducerPath]: githubApi.reducer,
	[usersGithubApi.reducerPath]: usersGithubApi.reducer,
});

export const store = configureStore({
	reducer: rootReducer,
	middleware: (getDefaultMiddleware) =>
		getDefaultMiddleware().concat(githubApi.middleware).concat(usersGithubApi.middleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;

setupListeners(store.dispatch);

*/