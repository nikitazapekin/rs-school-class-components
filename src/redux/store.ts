import { configureStore, combineReducers } from "@reduxjs/toolkit";
import postsSlice from "./slices/postsSlice";
import type { ThunkAction, Action } from '@reduxjs/toolkit';
import appSlice from "./slices/appSlice";
import selectedElementsSlice from "./slices/selectedElementsSlice";
import { githubApi } from "./slices/querySlice";
import { usersGithubApi } from "./slices/userQuerySlice";
const rootReducer = combineReducers({
    appSlice: appSlice,
    selectedElementsSlice: selectedElementsSlice,
    postsSlice: postsSlice,
    [githubApi.reducerPath]: githubApi.reducer,
    [usersGithubApi.reducerPath]: usersGithubApi.reducer,
    posts: postsSlice
});
export const store = configureStore({
    //  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware().concat(githubApi.middleware).concat(usersGithubApi.middleware),
    //  });
    //  reducer: {
    //    posts: postsSlice
    // }
})



export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
/*
export type AppStore = ReturnType<typeof store>;
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<AppStore['getState']>;
export type AppDispatch = AppStore['dispatch'];
*/
/*
export const store = configureStore({
    reducer: {
        posts: postsSlice
    }
})
*/



/*
//lib/store.ts

import { combineReducers, configureStore } from '@reduxjs/toolkit';
import type { ThunkAction, Action } from '@reduxjs/toolkit';
import appSlice from './slices/appSlice';
import { setupListeners } from '@reduxjs/toolkit/query';
import { githubApi } from './slices/querySlice';
import { usersGithubApi } from './slices/userQuerySlice';
import selectedElementsSlice from './slices/selectedElementsSlice';
import testSlice from './slices/testSlice';

const rootReducer = combineReducers({
  appSlice: appSlice,
  selectedElementsSlice: selectedElementsSlice,
  testSlice: testSlice,
  [githubApi.reducerPath]: githubApi.reducer,
  [usersGithubApi.reducerPath]: usersGithubApi.reducer,
});

export const makeStore = () => {
  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(githubApi.middleware).concat(usersGithubApi.middleware),
  });
};

// Infer the type of makeStore
export type AppStore = ReturnType<typeof makeStore>;
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<AppStore['getState']>;
export type AppDispatch = AppStore['dispatch'];

// Set up listeners
const store = makeStore();
setupListeners(store.dispatch);

export default store;

*/