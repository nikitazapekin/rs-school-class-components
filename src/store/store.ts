import { combineReducers, configureStore } from "@reduxjs/toolkit";
import type { ThunkAction, Action } from '@reduxjs/toolkit';
import appSlice from "./slices/appSlice";
import { setupListeners } from '@reduxjs/toolkit/query'
import { githubApi } from "./slices/querySlice";
import { usersGithubApi } from "./slices/userQuerySlice";

export const rootReducer = combineReducers({
  appSlice: appSlice,
  [githubApi.reducerPath]: githubApi.reducer,
  [usersGithubApi.reducerPath]: usersGithubApi.reducer,  
});

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware()
      .concat(githubApi.middleware)
      .concat(usersGithubApi.middleware),  
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;

setupListeners(store.dispatch);



 