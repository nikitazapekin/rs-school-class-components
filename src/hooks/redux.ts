import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
//import { AppDispatch, RootState } from '@/store/store';
import { RootState, AppDispatch } from '../lib/store';
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
