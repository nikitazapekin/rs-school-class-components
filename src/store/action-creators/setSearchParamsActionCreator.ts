import { setSearchParamsURL, setNextPage, setPrevPage, setQueryPage, setFirstPage } from "../slices/appSlice"
import { AppDispatch } from "../store"
export const setSearchParamsActionCreator = (offset: number, query: string) => (dispatch: AppDispatch) => {
    dispatch(setSearchParamsURL({ offset, query }))
}
export const setFirstPageActionCreator = () => (dispatch: AppDispatch) => {
    dispatch(setFirstPage())
}
export const setNextPageActionCreator = () => (dispatch: AppDispatch) => {
    dispatch(setNextPage())
}
export const setPrevPageActionCreator = () => (dispatch: AppDispatch) => {
    dispatch(setPrevPage())
}
export const setQueryActionCreator = (query: string) => (dispatch: AppDispatch) => {
   dispatch(setQueryPage(query))
}