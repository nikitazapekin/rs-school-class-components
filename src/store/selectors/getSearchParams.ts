import { RootState } from "../store";
export const paramsSelector = (state: RootState) => state.appSlice.params