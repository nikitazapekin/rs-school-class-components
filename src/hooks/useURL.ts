import { useState, useEffect } from "react";
import { useSearchParams } from "react-router-dom";
import { paramsSelector } from "@/store/selectors/getSearchParams";
import { useNavigate } from "react-router-dom";
import { useAppDispatch } from "./redux"
import { useSelector } from "react-redux";
import { setQueryActionCreator } from "@/store/action-creators/setSearchParamsActionCreator";
import { setSearchParamsActionCreator } from "@/store/action-creators/setSearchParamsActionCreator";
import { setLoadingActionCreator } from "@/store/action-creators/setIsLoading";
import { useLazySearchUsersQuery } from "@/store/slices/querySlice";
import { setUsersActionCreator } from "@/store/action-creators/setUsersActionCreator";
const useURL = () => {
    const dispatch = useAppDispatch()
    const [searchParams, setSearchParams] = useSearchParams();
    const getCurrentParams = () => {
        const page = parseInt(searchParams.get('page') || '1', 10);
        const query = searchParams.get('query') || '';
        return [page, query];
    };

    const setPage = (page: number, query: string) => {
        const params: Record<string, string> = { page: String(page) };
        if (query.length > 0) {
            params.query = query;
        }
        setSearchParams(params);
    };
    useEffect(() => {
        const [page, query] = getCurrentParams();
        setPage(Number(page), String(query));
        localStorage.setItem('searchParam', String(query));
        dispatch(setSearchParamsActionCreator(Number(page), String(query)))

    }, []);
    const navigate = useNavigate();
    const params = useSelector(paramsSelector)
    const handleRedirect = () => {
        navigate('/not-existing-page');
    };
    const handleInput = (event: React.ChangeEvent<HTMLInputElement>) => {
        dispatch(setQueryActionCreator(event.target.value))

    };
    const [trigger, { data, error, isLoading }] = useLazySearchUsersQuery();
    const handleSearch = () => {
        window.scrollTo(0, 0);
        trigger({ query: params.query, page: params.offset, per_page: params.limit });
        localStorage.setItem('searchParam', params.query)
        setPage(params.offset, params.query)
        dispatch(setSearchParamsActionCreator(1, params.query))
    };
    useEffect(() => {
        if (isLoading) {
            dispatch(setLoadingActionCreator(true))
        } else {
            dispatch(setLoadingActionCreator(false))
        }
        if (data) {
            dispatch(setUsersActionCreator(data.items))
        }
    }, [data]);
    useEffect(() => {
        trigger({ query: params.storedValue ? params.storedValue : params.query, page: params.offset, per_page: params.limit });
        setPage(params.offset, params.storedValue)
    }, [params.storedValue])
    useEffect(() => {
        trigger({ query: params.query, page: params.offset, per_page: params.limit });
        setPage(params.offset, params.query)
    }, [params.offset])
    return { getCurrentParams, setPage, handleInput, handleSearch, handleRedirect }
}
export default useURL