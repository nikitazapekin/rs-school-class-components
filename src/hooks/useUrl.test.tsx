import { renderHook, act } from '@testing-library/react-hooks';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { useNavigate, useLocation } from '@remix-run/react';
import { useAppDispatch } from './redux';
import { paramsSelector } from '../store/selectors/getSearchParams';
import { setSearchParamsActionCreator } from '../store/action-creators/setSearchParamsActionCreator';
import { setQueryActionCreator } from '../store/action-creators/setSearchParamsActionCreator';
import useURL from './useURL';
import { rootReducer } from '../store/store';
import themeSlice from '@/store/slices/themeSlice';

jest.mock('@remix-run/react', () => ({
  useNavigate: jest.fn(),
  useLocation: jest.fn(),
}));

jest.mock('./redux', () => ({
  useAppDispatch: jest.fn(),
}));

jest.mock('../store/selectors/getSearchParams', () => ({
  paramsSelector: jest.fn(),
}));

jest.mock('../store/action-creators/setSearchParamsActionCreator', () => ({
  setSearchParamsActionCreator: jest.fn(),
  setQueryActionCreator: jest.fn(),
}));

const mockDispatch = jest.fn();
const mockNavigate = jest.fn();
const mockLocation = {
  search: '?page=1&query=test'
};

const initialState = {
    selectedElementsSlice: {
        storedElements: [],

	selectedElement: null,
    },
  appSlice: {
    isLoading: false,
	error: null,
	isLoadingUserData: false,
	users: [],
	typedValue: "",
	params: {
		limit: 10,
		offset: 1,
		query: '',
		storedValue: '',
	},
	status: 'idle',
	clickedUser: {
		login: "",
	id: 0,
	node_id:"" ,
	avatar_url: "",
	gravatar_id:"" ,
	url:"" ,
	html_url:"" ,
	followers_url: "",
	following_url:"" ,
	gists_url: "",
	starred_url: "",
	subscriptions_url:  "",
	organizations_url: "",
	repos_url:"" ,
	events_url:"" ,
	received_events_url:"" ,
	type: "",
	site_admin: false,
	score: 0
	}
  },
  themeSlice: {
    isDark: false
  }
};

const createTestStore = (state: typeof initialState) =>
  configureStore({
    reducer: rootReducer,
    preloadedState: state,
  });

describe('useURL hook', () => {
  beforeEach(() => {
    (useAppDispatch as jest.Mock).mockReturnValue(mockDispatch);
    (useNavigate as jest.Mock).mockReturnValue(mockNavigate);
    (useLocation as jest.Mock).mockReturnValue(mockLocation);
    (paramsSelector as jest.Mock).mockReturnValue(initialState.appSlice.params);
  });

  test('should set page and query in URL', () => {
    const store = createTestStore(initialState);
    const wrapper = ({ children }: { children: React.ReactNode }) => <Provider store={store}>{children}</Provider>;

    const { result } = renderHook(() => useURL(), { wrapper });

    act(() => {
      result.current.setPage(2, 'newQuery');
    });

    expect(mockNavigate).toHaveBeenCalledWith('?page=2&query=newQuery');
  });

  test('should dispatch setSearchParamsActionCreator on mount', () => {
    const store = createTestStore(initialState);
    const wrapper = ({ children }: { children: React.ReactNode }) => <Provider store={store}>{children}</Provider>;

    renderHook(() => useURL(), { wrapper });

    expect(mockDispatch).toHaveBeenCalledWith(setSearchParamsActionCreator(1, 'test'));
  });

  test('should handle input change and dispatch setQueryActionCreator', () => {
    const store = createTestStore(initialState);
    const wrapper = ({ children }: { children: React.ReactNode }) => <Provider store={store}>{children}</Provider>;

    const { result } = renderHook(() => useURL(), { wrapper });

    act(() => {
      result.current.handleInput({ target: { value: 'newQuery' } } as React.ChangeEvent<HTMLInputElement>);
    });

    expect(mockDispatch).toHaveBeenCalledWith(setQueryActionCreator('newQuery'));
  });

  test('should handle search and navigate with new query', () => {
    const store = createTestStore(initialState);
    const wrapper = ({ children }: { children: React.ReactNode }) => <Provider store={store}>{children}</Provider>;

    const { result } = renderHook(() => useURL(), { wrapper });

    act(() => {
      result.current.handleSearch();
    });

    expect(mockDispatch).toHaveBeenCalledWith(setSearchParamsActionCreator(1, 'test'));
    expect(mockNavigate).toHaveBeenCalledWith('?page=1&query=test');
  });
});
