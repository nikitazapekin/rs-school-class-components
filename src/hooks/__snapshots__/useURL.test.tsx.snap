// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useURL should match snapshot 1`] = `
{
  "all": [
    {
      "getCurrentParams": [Function],
      "handleInput": [Function],
      "handleRedirect": [Function],
      "handleSearch": [Function],
      "setPage": [Function],
    },
    {
      "getCurrentParams": [Function],
      "handleInput": [Function],
      "handleRedirect": [Function],
      "handleSearch": [Function],
      "setPage": [Function],
    },
    {
      "getCurrentParams": [Function],
      "handleInput": [Function],
      "handleRedirect": [Function],
      "handleSearch": [Function],
      "setPage": [Function],
    },
    {
      "getCurrentParams": [Function],
      "handleInput": [Function],
      "handleRedirect": [Function],
      "handleSearch": [Function],
      "setPage": [Function],
    },
  ],
  "current": {
    "getCurrentParams": [Function],
    "handleInput": [Function],
    "handleRedirect": [Function],
    "handleSearch": [Function],
    "setPage": [Function],
  },
  "error": undefined,
}
`;
